### PR TITLE
Add linter pre-commit hook 

### PR DESCRIPTION
**Pull request by tlambert03**
# Description
picking up the conversation started in https://github.com/napari/napari/pull/633#issuecomment-546956411, this PR would add a pre-commit lint hook to prevent things like syntax or logical errors from being committed.

This PR is just a stub for now, to give everyone a chance to play with various linters.  My (very weak) personal preference has been for flake8.  While it's true that it covers a bit less than pylint I find that it's got more than enough to prevent me from accidentally introducing bugs, and can be expanded with plugins.  Ultimately, I found that if I just wanted to prevent actual errors from creeping in, and assert a little more conventions, flake8 was sufficient... whereas pylint took more time to tune so as not to get in the way.  On the flip side, there's obviously more momentum behind pylint.

I'm definitely not attached to flake8, and it's your call!  if you come to have a preference one way or the other, feel free to directly edit my branch, or leave a comment.

## Type of change
development environment

# References
https://github.com/PyCQA/pylint
https://github.com/PyCQA/flake8

## Final checklist:
- [x] add precommit hooks
- [x] agree upon linter configuration and put settings in `setup.cfg`
- [x] (optional) - run the linter on the entire repo and make sure it passes, so that we don't have a lot of small unrelated changes each time someone edits a file.  This has the potential to cause a lot of merge conflicts in the short term, but will help prevent commit failures in the long term.

### CODE REVIEW

**Comment by tlambert03** [comment]
After playing a bit more with both flake8 and pylint, I decided to just push forward with flake8.  It was faster than pylint and should prevent any logical bugs from slipping in as well as many stylistic ones.  If in the future we'd like to move to pylint, it should be straightforward.

These are the errors I chose to ignore for now. 
See https://lintlyci.github.io/Flake8Rules for details
```
# E203  Whitespace before ':'  (sometimes conflicts with black)
# E501  line too long (> 79 characters)  (sometimes too annoying)
# W503	Line break occurred before a binary operator (conflicts with black)
# E711	Comparison to none should be 'if cond is none:' (there are tons of these)
# E712	Comparison to true should be 'if cond is true:' or 'if cond:' (lots)
# C901  McCabe complexity test.
```
The two main offenders of the [complexity test](https://lintlyci.github.io/Flake8Rules/rules/C901.html) were `Shapes._move` and `Shapes.on_mouse_press`, so if we can simplify those, it might be nice to re-enable that.

To bring the whole repo into flake8 compatibility with the remaining checks (e.g. `flake8 napari` at the command line), I fixed the following errors.  Note, many of these were fixed simply by running black... which would have happened eventually.

```
E114 indentation is not a multiple of four (comment)
E116 unexpected indentation (comment)
E123 closing bracket does not match indentation of opening bracket's line
E126 continuation line over-indented for hanging indent
E127 continuation line over-indented for visual indent
E131 continuation line unaligned for hanging indent
E123 closing bracket does not match indentation of opening bracket's line
E201 whitespace after '['
E202 whitespace before ']'
E241 multiple spaces after ':'
E302 expected 2 blank lines
E305 expected 2 blank lines after class or function definition
E704 multiple statements on one line (def)
F841 local variable assigned to but never used
W293 blank line contains whitespace
W504 line break after binary operator
```

I'm also currently ignoring all `__init__.py` files, since they have a lot of unused imports.  The "recommended" way to deal with those seems to be either to add a `# noqa` comment:
```python
from .my_class import MyClass  # noqa
```
or to define `__all__`
```python
from .my_class import MyClass
__all__ = ['MyClass',]
```
but it's up to you whether you'd like to enable that, or avoid the boilerplate.
This is all set for review/feedback.

File: .pre-commit-config.yaml
Comment (bjlittle) [inline .pre-commit-config.yaml]: @tlambert03 `3.7.9` just landed a few days ago

File: .pre-commit-config.yaml
Comment (bjlittle) [inline .pre-commit-config.yaml:13]: @tlambert03 Apologies, I'm pretty new to pre-commit hooks... but isn't the `pass_filenames` by default `true`, see [here-ish](https://pre-commit.com/#new-hooks).

If that's the case, then is there a reason to set this explicitly? Just curious, thanks :smile:

Comment (tlambert03) [reply .pre-commit-config.yaml:13]: I thought being explicit would make it easier for people to quickly glance at the config and know what was going on (without searching the docs)

Comment (tlambert03) [reply .pre-commit-config.yaml]: I thought being explicit would make it easier for people to quickly glance at the config and know what was going on (without searching the docs)

Comment (bjlittle) [reply .pre-commit-config.yaml:13]: Nice one, thanks @tlambert03 

Makes sense to me. I just wanted to clarify (for my understanding) that's what you were doing and not working around some other issue. Awesome!

**Comment by bjlittle** [comment]
@tlambert03 At the moment I'm attempting to `black`ify and `flake8` the `iris` code base, see https://github.com/SciTools/iris/pull/3502/files. So I'm facing similar challenges.

I've also opted to ignore a handful of warnings and errors, which when compared with other scientific python packages, seems reasonable; particularly since `black` is involved.

IMHO the warnings and errors that you're ignoring here in this PR seems totally reasonable to me. However, I'd suggest that we should aim to resolve the `E711` and `E712` in a separate single follow on PR. I'm guessing that to resolve  `C901` may be harder and require more  concerted effort, so as a strategy going forward I'd suggest also resolving that separately.

If you buy into that approach, then perhaps raise associated issues to cover those two pieces of work i.e., `E711`+`E712` and `C901`, so that they are not forgotten about, and we just bank the goodness of this PR "as is"?

**Comment by tlambert03** [comment]
Hi @bjlittle, it would be easy enough to fix `E711`&`E712` in this PR.  I was mostly waiting to give @jni and @sofroniewn an opportunity to look at the existing changes and weigh in before editing a lot more lines of the existing code base.

File: napari/_qt/qt_main_window.py
Comment (sofroniewn) [inline napari/_qt/qt_main_window.py]: why do we end up with a `# noqa` here?

**Review by sofroniewn — COMMENTED** [review]
Looking over it, it seems ok to me - though we had been ignoring the `vendored` colormaps stuff I think for `black` so it was kept the same as the code from other places, right @jni? Might want to ignore here too.

I also don't really like the appearance of `# noqa`. To me they do more harm then good (confuse a beginner, whereas only provide minimal clarity to an expert). Can we disable them?

In general though I don't have super strong opinions about this stuff (though I like having as much automated as possible) and defer to @jni.

Also your comment about the complexity of some code in the shapes layer makes sense, we should probably improve that, but that can be done in new PRs

File: napari/tests/test_viewer.py
Comment (bjlittle) [inline napari/tests/test_viewer.py]: @tlambert03 Can you drop the `# noqa` here and just have:
```python
with pytest.raises(RuntimeError):
    Viewer()
```

Comment (jni) [reply napari/tests/test_viewer.py]: Another alternative with me hoping against hope: does `_ = Viewer()` get ignored?

File: napari/_qt/qt_main_window.py
Comment (jni) [inline napari/_qt/qt_main_window.py]: What is the "violation" here? (and in the line below)

File: napari/_qt/qt_viewer.py
Comment (jni) [inline napari/_qt/qt_viewer.py:268]: Why is this change needed? (sorry to be "nosy" — I want to know for my own understanding but also to decide whether I want to add this to the ignore list...)

File: napari/components/tests/test_viewer_model.py
Comment (jni) [inline napari/components/tests/test_viewer_model.py]: Yeah, these are annoying to me, mainly because `*_layer = viewer.add_*(data)` is a "canonical" form for napari so we would like to keep to it as much as possible. Perhaps we can test some aspect of the added layer, e.g. `assert image_layer.dims.ndim == 3`?

File: napari/layers/points/tests/test_points.py
Comment (jni) [inline napari/layers/points/tests/test_points.py:49]: nice one! :sweat_smile:

**Comment by tlambert03** [comment]
> though we had been ignoring the vendored colormaps stuff I think for black so it was kept the same as the code from other places, right @jni? Might want to ignore here too.

sorry, i missed that in pyproject... I'll ignore it in linting too.

> I also don't really like the appearance of # noqa. To me they do more harm then good (confuse a beginner, whereas only provide minimal clarity to an expert). Can we disable them?

yeah I hear ya... We can't disable them per se.  If you have code that doesn't follow one of the linting rules, you can either 
1.  ignore the specific error globally 
2. ignore the entire file
3. ignore the offending line with `# noqa`
http://flake8.pycqa.org/en/3.1.1/user/ignoring-errors.html

So every so often you come across something you want to ignore locally, but not globally or across the whole file.  One option is to agree to be more specific about the comment, e.g. `# noqa: E731,E123`, which provides more clarity to experts, though still probably confuses beginners. Fixing the error completely is preferable, but not always possible (though the few `noqa's` here are not necessarily that case, I was just being "gentle" with corrections).

File: napari/_qt/qt_main_window.py
Comment (tlambert03) [reply napari/_qt/qt_main_window.py]: that code breaks  [Module level import not at top of file (E402) ](https://lintlyci.github.io/Flake8Rules/rules/E402.html) ... but I didn't want to just move the imports, since that can sometimes alter behavior.  so rather than globally ignoring `E402`, I copped out with the  `# noqa`.  see my comment below about `#noqa` ... but I can also take a closer look to make sure moving the imports wouldn't break anything here.

File: napari/util/colormaps/vendored/_color_data.py
Comment (jni) [inline napari/util/colormaps/vendored/_color_data.py]: Yeah, @sofroniewn is right: let's ignore everything in `vendored/`

File: napari/tests/test_viewer.py
Comment (tlambert03) [reply napari/tests/test_viewer.py]: @bjlittle, yes you could.  But i was again opting for the more explicit option there, since @jni wrote that code and I didn't risk messing up the test.  @jni, you're in luck: flake8 will ignore unused underscore variables.  I think that's the best solution here.

File: napari/_qt/qt_main_window.py
Comment (tlambert03) [reply napari/_qt/qt_main_window.py]: same as above: [E402](https://lintlyci.github.io/Flake8Rules/rules/E402.html).  the problem is the `app.use_app(API_NAME)` above it.  options are to ignore E402 globally, skip this file, or use `# noqa`

File: napari/util/tests/test_keybindings.py
Comment (jni) [inline napari/util/tests/test_keybindings.py]: Yeah, like @sofroniewn, I'm a little allergic to the `# noqa`'s everywhere. If there is a way to configure this to instead be `#flake8ignore` or somesuch (explicit is better than implicit), I'd be more inclined to ignore (heh) them. As it is, I think it's better to do as it says, and replace this with an explicit function definition.

File: napari/_qt/qt_main_window.py
Comment (bjlittle) [reply napari/_qt/qt_main_window.py]: @jni I'm guessing that it's caused by [lines 12-13](https://github.com/napari/napari/blob/master/napari/_qt/qt_main_window.py#L12-L13) coming between the `import` statements.

I don't know `napari` well enough yet (working on that, ♥️ it btw), but could:
```python
app.use_app(API_NAME)
del app
```
be moved after all the `import` statements?

File: setup.cfg
Comment (jni) [inline setup.cfg:14]: Never heard of this one! I quite like it! Yes, I envision making multiple (many?) smaller PRs to get this fixed. I wouldn't look forward to reviewing a huge one that tackles it for the whole library!

File: setup.cfg
Comment (jni) [inline setup.cfg]: Good call waiting on the first round of review to fix these, but I agree with @bjlittle (:wave:!) that we should definitely tackle them!

**Review by jni — CHANGES_REQUESTED** [review]
@tlambert03 thanks! Man, enabling a linter takes work! :sweat_smile: I've left a few comments, but overall I'm pretty happy with how this is shaping up!

File: napari/_qt/qt_main_window.py
Comment (jni) [reply napari/_qt/qt_main_window.py]: @bjlittle :tada: so great to see you here. =D If I remember correctly, it needs to come before the PyQt imports, but I'm not positive about this. The basic idea is to get VisPy and QtPy speaking the same "dialect" of Qt. Trying to figure this out for our [conda-forge recipe](https://github.com/conda-forge/staged-recipes/pull/9983), in case you might have input there! :grimacing:

File: napari/_qt/qt_viewer.py
Comment (tlambert03) [reply napari/_qt/qt_viewer.py:268]: lol @ the concept of _you_ being "nosy" on a napari PR.
The issue there was `E501` (line too long).  Sometimes, black is unable to come up with a satisfying solution (like in that case with a few long chained attributes).  Black will fail gracefully but flake8 won't, so that was just a way to decrease the line length.

File: napari/util/tests/test_keybindings.py
Comment (jni) [reply napari/util/tests/test_keybindings.py]: @tlambert03 I definitely prefer the more explicit `#noqa: E<whatever>` form, thanks for pointing that out. It's a strong hint about e.g. what to Google to figure out what that comment is doing.

File: napari/tests/test_viewer.py
Comment (jni) [reply napari/tests/test_viewer.py]: YASSSS! Let's do that where needed!

File: napari/_qt/qt_viewer.py
Comment (jni) [reply napari/_qt/qt_viewer.py:268]: Ah good! It hadn't occurred to me that black could be unable to fix that. Happy with this workaround. Thanks!

Comment (jni) [reply napari/_qt/qt_viewer.py:268]: Also: I try to be mindful to minimize annoying nitpicks in PRs, but sometimes I can't help myself! :joy:

File: napari/components/tests/test_viewer_model.py
Comment (tlambert03) [reply napari/components/tests/test_viewer_model.py]: yeah, i guess these are the annoyances of linting 😕
how do you feel about `_ = viewer.add_image(image)` here?  It's just a test after all (but I recognize that tests are a great place for people to learn to use it, and we'd like to establish conventions there)

Comment (jni) [reply napari/components/tests/test_viewer_model.py]: Yep, let's go for it! Indicating at least that something is returned is good enough for me!

File: napari/_qt/qt_main_window.py
Comment (sofroniewn) [reply napari/_qt/qt_main_window.py]: those imports might need to be there, I can't fully remember, but I do think there were a couple places like that due to qt stuff

File: napari/_qt/qt_main_window.py
Comment (sofroniewn) [reply napari/_qt/qt_main_window.py]: that is my memory too @jni - i don't think it can be moved

**Comment by tlambert03** [comment]
Just getting off train, will address all of these and resubmit soon

File: setup.cfg
Comment (tlambert03) [reply setup.cfg:14]: if by "this one" you mean the mccabe test, I think we're actually in amazingly good shape that there are only two overly complex methods (on `Shapes`).  probably won't take long for someone to split those into a few methods and pass that test.

File: napari/util/tests/test_keybindings.py
Comment (tlambert03) [reply napari/util/tests/test_keybindings.py]: looked into briefly but could not find any way to change the syntax anything other than `# noqa`.... will opt for explicit error codes

**Comment by tlambert03** [comment]
- undid all changes within `vendored` and added it to `excludes` (note, the excludes are duplicated in `setup.cfg` and `pre-commit-config.yaml` ... I found some threads a similar requirement and went with it).
- fixed all `E711` and `E712`, (Comparison to `None`/`True`/`False`) and removed corresponding ignores from `setup.cfg`
- minimized `noqa`s.  All remaining ones have explicit flake8 codes
- ready to ignore `E501` (all lines ≥79 are either fixed or have `# noqa: E501`), but `E501` remains in `setup.cfg` ignores until everyone is ready for that level of linting strictness.

File: napari/util/tests/test_keybindings.py
Comment (bjlittle) [reply napari/util/tests/test_keybindings.py]: @tlambert03 You should be able to tell `flake8` to ignore `E731` to allow assignment of `lambda`s... could you give that a try? If that works, then you can drop the `# noqa`

Comment (tlambert03) [reply napari/util/tests/test_keybindings.py]: I already replaced it with an explicit function definition as @jni suggested above.  I don’t personally feel that assignment to lambdas provides any particular additional clarity besides simply fitting on one line... so hesitant to add E731 to ignores

**Review by sofroniewn — CHANGES_REQUESTED** [review]
Thanks for the changes - they all look good to me, but I'll let @jni weigh in for final approval

Do we have to do anything extra to enable the pre-commit hook (especially if we already have napari installed)? Can you maybe leave a comment on this PR containing the info for current devs.

Can you add a line or two about our linting to our contributing guide and possibly include some helpful links - it can go next to the part about setting up `black` here - https://github.com/napari/napari/blob/master/docs/CONTRIBUTING.md#setting-up-a-development-installation

File: napari/layers/image/tests/test_image.py
Comment (jni) [inline napari/layers/image/tests/test_image.py:209]: _ = 
```suggestion
        _ = Image(data, rgb=True)
```

Comment (jni) [reply napari/layers/image/tests/test_image.py:209]: Eh, ignore me, just saw that it's not supposed to even get to the assignment step! :joy:

File: napari/util/list/_typed.py
Comment (jni) [inline napari/util/list/_typed.py:10]: This violates PEP257 (one-liner should be one line :joy:), but we can leave that for a later PR, as there's lots of those to clean up.

**Review by jni — APPROVED** [review]

**Comment by tlambert03** [comment]
> Do we have to do anything extra to enable the pre-commit hook (especially if we already have napari installed)? Can you maybe leave a comment on this PR containing the info for current devs.

Assuming one has already installed pre-commit with `pre-commit install` as described in [CONTRIBUTING.md](https://github.com/napari/napari/blob/master/docs/CONTRIBUTING.md), the hook will be enabled once it gets merged, (but you may need to `pip install -r requirements.txt` if you don't already have flake8 installed)

> Can you add a line or two about our linting to our contributing guide and possibly include some helpful links - it can go next to the part about setting up `black` here - https://github.com/napari/napari/blob/master/docs/CONTRIBUTING.md#setting-up-a-development-installation

done!

**Comment by jni** [comment]
@sofroniewn all clear from me, ball's in your court! =)

File: napari/util/list/_typed.py
Comment (tlambert03) [reply napari/util/list/_typed.py:10]: yeah, that's a bit of a pickle isn't it!

**Review by sofroniewn — APPROVED** [review]
Looks great, thanks for adding those links. I will merge now

**Pull request merged**
